common/headtracker/head_tracker.go

```diff
@@ -302,19 +312,39 @@ func (ht *HeadTracker[HTH, S, ID, BLOCK_HASH]) backfillLoop() {
 	}
 }
 
+// calculateLatestFinalized - returns latest finalized block. It's expected that currentHeadNumber - is the head of
+// canonical chain. There is no guaranties that returned block belongs to the canonical chain. Additional verification
+// must be performed before usage.
+func (ht *HeadTracker[HTH, S, ID, BLOCK_HASH]) calculateLatestFinalized(ctx context.Context, currentHeadNumber int64) (h HTH, err error) {
+	if ht.config.FinalityTagEnabled() {
+		return ht.client.LatestFinalizedBlock(ctx)
+	}
+	finalizedBlockNumber := currentHeadNumber - int64(ht.config.FinalityDepth())
```

Comment:
As an optimization we can check if FinalityDepth is 0 and if it is then we can just return the initial head, avoiding making an extra RPC call. This is especially important on fast chains with instant finality.